# Effectiveness of Random Oversampling

In this activity, you’ll fit logistic regression models to both imbalanced data and resampled data. You’ll then compare the results by using the metrics that you’ve learned.

## Instructions

1. Read in the CSV file from the `Resources` folder into a Pandas DataFrame.  

2. Create a Series named `y` that contains the data from the "Default" column of the original DataFrame. Note that this Series will contain the labels. Create a new DataFrame named `X` that contains the remaining columns from the original DataFrame. Note that this DataFrame will contain the features.

3. Split the features and labels into training and testing sets.

4. Check the magnitude of imbalance in the data set by viewing  the number of distinct values  (`value_counts`) for the labels. 

5. Resample the training data by using `RandomOverSampler`.

6. Check the number of distinct values (`value_counts`) for the resampled labels.

7. Fit two logistic regression modules: one for the resampled data and another for the original data.

 8.  Using the two logistic regression models, predict the values for the original and resampled sets.

9. Print the confusion matrixes, accuracy scores, and classification reports for the original and resampled datasets.

10. Evaluate the effectiveness of random oversampling for predicting the minority class. Answer the following question: Does the model accurately flag all the loans that eventually defaulted?


## References

Following are links to modules from the scikit learn library that will be utilized:

[LogisticRegression](https://scikit-learn.org/stable/modules/generated/sklearn.linear_model.LogisticRegression.html)

[train_test_split](https://scikit-learn.org/stable/modules/generated/sklearn.model_selection.train_test_split.html)

[confusion_matrix](https://scikit-learn.org/stable/modules/generated/sklearn.metrics.confusion_matrix.html)

[balanced_accuracy_score](https://scikit-learn.org/stable/modules/generated/sklearn.metrics.balanced_accuracy_score.html)

Following are links to modules from the imbalanced learn library that will be utilized:

[RandomOverSampler](https://imbalanced-learn.org/stable/generated/imblearn.over_sampling.RandomOverSampler.html)

[classifiction_report_imbalanced](https://imbalanced-learn.org/stable/generated/imblearn.metrics.classification_report_imbalanced.html)

In [30]:
import pandas as pd
import matplotlib.pyplot as plt
from pathlib import Path

from sklearn.model_selection import train_test_split
from sklearn.linear_model import LogisticRegression
from sklearn.metrics import confusion_matrix, classification_report
from sklearn.metrics import balanced_accuracy_score

#from imblearn.over_sampling import RandomOverSampler
#from imblearn.metrics import classification_report_imbalanced

## Step 1: Read in the CSV file from the `Resources` folder into a Pandas DataFrame. 

In [13]:
# Read the sba_loans.csv file from the Resources folder into a Pandas DataFrame
loans_df = pd.read_csv(
    Path('12.4/01-Effectiveness_of_Random_Oversampling/Resources/sba_loans.csv')
)

# Review the DataFrame
loans_df.head()

,Year,Month,Amount,Term,Zip,CreateJob,NoEmp,RealEstate,RevLineCr,UrbanRural,Default
0,2001,11,32812,36,92801,0,1,0,1,0,0
1,2001,4,30000,56,90505,0,1,0,1,0,0
2,2001,4,30000,36,92103,0,10,0,1,0,0
3,2003,10,50000,36,92108,0,6,0,1,0,0
4,2006,7,343000,240,91345,3,65,1,0,2,0


## Step 2: Create a Series named `y` that contains the data from the "Default" column of the original DataFrame. Note that this Series will contain the labels. Create a new DataFrame named `X` that contains the remaining columns from the original DataFrame. Note that this DataFrame will contain the features.

In [14]:
# Split the data into X (features) and y (lables)

# The y variable should focus on the Default column
y = loans_df['Default']

X = loans_df.drop(columns=['Default'])
X.head()

,Year,Month,Amount,Term,Zip,CreateJob,NoEmp,RealEstate,RevLineCr,UrbanRural
0,2001,11,32812,36,92801,0,1,0,1,0
1,2001,4,30000,56,90505,0,1,0,1,0
2,2001,4,30000,36,92103,0,10,0,1,0
3,2003,10,50000,36,92108,0,6,0,1,0
4,2006,7,343000,240,91345,3,65,1,0,2


### Step 3: Split the features and labels into training and testing sets.

In [15]:
X_train,X_test,y_train,y_test = train_test_split(X,y,random_state=1)

## Step 4: Check the magnitude of imbalance in the data set by viewing  the number of distinct values  (`value_counts`) for the labels. 

In [16]:
# Count the distinct values in the orignal labels data
y_train.value_counts()

Default
0    1063
1      96
Name: count, dtype: int64

### Imbalance data in sklearn

In [7]:
LogisticRegression(class_weight="balanced")

LogisticRegression(class_weight='balanced')

In [9]:
# use 1063//96 to get a ratio of a non_fraud/fraud, 
# we know the data is 1063//96 ~ 11 times imbalanced
# this number is more practical than the auto balanced calculation
# note class_weight not create fake data like imbalanced, work well for linear model
# package like imbalanced works better for non linear model like tree-model or neuralnetwork
LogisticRegression(class_weight={0:1, 1: (1063//96)})

LogisticRegression(class_weight={0: 1, 1: 11})

## Step 5: Resample the training data by using `RandomOverSampler`.

In [17]:
display(X_train[:2])
display(X_test[:2])
display(y_train[:2])
display(y_test[:2])

,Year,Month,Amount,Term,Zip,CreateJob,NoEmp,RealEstate,RevLineCr,UrbanRural
298,2007,7,248000,240,92618,7,10,1,0,2
1415,2005,8,150000,53,93446,0,5,0,0,1


,Year,Month,Amount,Term,Zip,CreateJob,NoEmp,RealEstate,RevLineCr,UrbanRural
942,2003,1,5000,84,91351,0,1,0,0,2
614,2001,7,322400,240,93308,0,2,1,0,2


298     0
1415    1
Name: Default, dtype: int64

942    0
614    0
Name: Default, dtype: int64

## Step 7: Fit two logistic regression modules:
### Original LogisticRegression VS. class_weight='balanced'

In [18]:
# Declare a logistic regression model
# Set a random_state paramerter with a value of 1
lr_base_model = LogisticRegression(random_state=1)

In [19]:
lr_base = lr_base_model.fit(X_train,y_train)
lr_base

/opt/miniconda3/envs/Linlin/lib/python3.13/site-packages/sklearn/linear_model/_logistic.py:465: ConvergenceWarning: lbfgs failed to converge (status=1):
STOP: TOTAL NO. OF ITERATIONS REACHED LIMIT.

Increase the number of iterations (max_iter) or scale the data as shown in:
    https://scikit-learn.org/stable/modules/preprocessing.html
Please also refer to the documentation for alternative solver options:
    https://scikit-learn.org/stable/modules/linear_model.html#logistic-regression
  n_iter_i = _check_optimize_result(


LogisticRegression(random_state=1)

In [20]:
lr_balanced_model = LogisticRegression(class_weight='balanced')
lr_balanced = lr_balanced_model.fit(X_train,y_train)
lr_balanced

/opt/miniconda3/envs/Linlin/lib/python3.13/site-packages/sklearn/linear_model/_logistic.py:465: ConvergenceWarning: lbfgs failed to converge (status=1):
STOP: TOTAL NO. OF ITERATIONS REACHED LIMIT.

Increase the number of iterations (max_iter) or scale the data as shown in:
    https://scikit-learn.org/stable/modules/preprocessing.html
Please also refer to the documentation for alternative solver options:
    https://scikit-learn.org/stable/modules/linear_model.html#logistic-regression
  n_iter_i = _check_optimize_result(


LogisticRegression(class_weight='balanced')

## Step 8: Using the two logistic regression models, predict the values for the original and resampled sets.

In [21]:
y_base_pred = lr_base.predict(X_test)

In [22]:
y_balanced_pred = lr_balanced.predict(X_test)

## Step 9: Print the confusion matrixes, accuracy scores, and classification reports for the original and resampled datasets.

In [23]:
confusion_matrix(y_test,y_base_pred)

array([[343,   5],
       [ 32,   7]])

In [24]:
confusion_matrix(y_test,y_balanced_pred)

array([[261,  87],
       [  7,  32]])

In [25]:
# Print the accuracy score for the original data
base_accuracy = balanced_accuracy_score(y_test,y_base_pred)
base_accuracy

np.float64(0.5825596816976127)

In [26]:
balanced_accuracy = balanced_accuracy_score(y_test,y_balanced_pred)
balanced_accuracy

np.float64(0.7852564102564102)

In [33]:
# Print the classification report for the original data
print(classification_report(y_test,y_base_pred))

              precision    recall  f1-score   support

           0       0.91      0.99      0.95       348
           1       0.58      0.18      0.27        39

    accuracy                           0.90       387
   macro avg       0.75      0.58      0.61       387
weighted avg       0.88      0.90      0.88       387



In [34]:
print(classification_report(y_test,y_balanced_pred))

              precision    recall  f1-score   support

           0       0.97      0.75      0.85       348
           1       0.27      0.82      0.41        39

    accuracy                           0.76       387
   macro avg       0.62      0.79      0.63       387
weighted avg       0.90      0.76      0.80       387



## Step 10: Evaluate the effectiveness of random oversampling for predicting the minority class. Answer the following question.

**Question:** Does the model generated using the resampled data more accurately flag all the loans that eventually defaulted?
    
**Answer:** The results regarding accuracy of the minority class are actually mixed when comparing the classifiction reports generated from the predictions with the original data versus the predictions with the resampled data. 

First, the accuracy score is much higher for the resampled data (0.81 vs 0.56), meaning that the model using resampled data was much better at detecting true positives and true negatives. 

The precision for the minority class is higher with the orignal data (0.55) versus the resampled data (0.32) meaning that the original data was better at detecting the users that were actually going to default. 

In terms of the recall, however, the minority class metric using resampled data was much better (0.82 vs 0.15). Meaning that the resampled data correctly clasified a higher percentage of the truly defaulting borrowers. 

All in, the model using resampled data was much better at detecting borrowers who are likely to default that the model generated using the original, imbalanced dataset. 